mount/mount_darwin.go

```diff
@@ -0,0 +1,121 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/ukontainer/darwin-snapshotter/hdiu"
+)
+
+var (
+	// ErrNotImplementOnUnix is returned for methods that are not implemented
+	ErrNotImplementOnUnix = errors.New("not implemented under darwin")
+	// allowedHelperBinaries is a list of third-party executable, which can
+	// be extended in future uses, e.g., mount_nfs, mount_9p.  `mount_darwin`
+	// is a special case where we will use the internal darwin snapshotter.
+	allowedHelperBinaries = []string{"mount_darwin"}
+)
+
+// Mount to the provided target path.
+//
+func (m *Mount) Mount(target string) error {
+	var allowed = false
+	log.L.Debugf("mount: %s, src %s, type %s", target, m.Source, m.Type)
+
+	binary := fmt.Sprintf("mount_%s", m.Type)
+	if m.Type != "darwin" {
+		if _, err := exec.LookPath(binary); err != nil {
+			return fmt.Errorf("mount binary not found: '%s(%s)': %w", m.Type, binary, errdefs.ErrNotFound)
+		}
+	}
+
+	for _, helperBinary := range allowedHelperBinaries {
+		if binary == helperBinary {
+			allowed = true
+		}
+	}
+	if !allowed {
+		return fmt.Errorf("mount binary isn't allowed to use: '%s(%s)': %w", m.Type, binary, ErrNotImplementOnUnix)
+	}
+
+	switch m.Type {
+	case "darwin":
+		return hdiu.AttachSparseBundle(target, m.Source, m.Options)
```

Comment:
Can we execute hdiu as a separate binary?

Reply:
this is something I was worried about..
okay, i'll try to prepare a binary to do mount/unmount stuffs.

Reply:
fixed it with using an external binary (mount_containerd_darwin), provided by the proxy snapshotter (https://github.com/ukontainer/darwin-snapshotter/tree/main/cmd/mount_containerd_darwin).